### DISCUSSION TITLE
ELI5: why does this not work? :(

### DISCUSSION THREAD

**Discussion by ptrxyz** [colinhacks/zod#3188] [Q&A] [answered]
![image](https://github.com/colinhacks/zod/assets/209056/e5b57c77-f678-4022-b5f2-d1741ff5cd4d)

See the depicted code. Why would type inference not work when I pass a schema to a function?
Yes, I can explicitly define the return type `as z.infer<A>`  in my function, but I wonder _why_ it doesn't work with inference.

**Comment by JacobWeisenburger** [comment]
Please send a full reproducible code example.

https://www.youtube.com/shorts/xlTrDIHhnhs

**Reply by ptrxyz** [reply]
There you go:

```ts
import { z } from 'zod'

function parse<A extends z.ZodTypeAny>(schema: A, value: unknown) {
	return schema.parse(value)
}

const value = parse(z.number(), '123') // type is lost. value should be typed as a number, yet it is Typescript thinks it's any
```

I simply would like to understand why TS can't infer the return type of `schema.parse` when passed through the (generic) function.

**Comment by JacobWeisenburger** [comment, marked as answer]
In my experience with generic functions, you need to define the return type. Something like this:

```ts
function parse<A extends z.Schema> ( schema: A, value: unknown ): z.infer<A> {
    return schema.parse( value )
}

const value = parse( z.number(), '123' )
// const value: number
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
